File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/policies/TopUpPaymentRequestPolicy.java
package it.giannibombelli.wsc2026.giftcard.application.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.application.Policy;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardTopUpRequested;
import it.giannibombelli.wsc2026.payment.application.commands.RequestPayment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;


public class TopUpPaymentRequestPolicy implements Policy<GiftCardTopUpRequested, RequestPayment> {

    @Override
    public RequestPayment evaluate(GiftCardTopUpRequested event) {
        Require.requireArgument(event, "Top-up request event");
        return new RequestPayment(
            EntityId.generate(PaymentId::new),
            new ClientReference(event.aggregateId().value()),
            event.requestedAmount(),
            Timestamp.now()
        );
    }
}

File: typescript/src/giftcard/application/policies/topUpPaymentRequestPolicy.ts
// Policy che traduce GiftCardTopUpRequested in RequestPayment per il Payment BC.
// Violazione cross-BC didattica: GiftCard non dovrebbe importare tipi da Payment.

import type { Policy } from '@/common/application/policy.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { RequestPayment, requestPayment } from '@/payment/application/commands/requestPayment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { GiftCardTopUpRequested } from '../../domain/events/giftCardTopUpRequested.js';

export class TopUpPaymentRequestPolicy implements Policy<GiftCardTopUpRequested, RequestPayment> {
  evaluate(event: GiftCardTopUpRequested): RequestPayment {
    requireArgument(event, 'Top-up request event');
    return requestPayment(
      generateId((value) => new PaymentId(value)),
      new ClientReference(event.aggregateId.value),
      event.requestedAmount,
      Timestamp.now()
    );
  }
}
